Preliminary Aggregate Ops and optimiaztion for Homogeneous DataFrames

Kriyszig opened the pull request:
This PR adds some preliminary aggregate operations and reducing of traversal overhead when targeting a single column of DataFrame by replacing TypeTuple with an array. (Same thing done for Group as well)
Implemented some preliminary aggregate operations.

cc/ @thewilsonator, ready for review (^_^)

thewilsonator commented on source/magpie/group.d:
```diff
@@ -175,7 +178,18 @@ public:
+        static if(isHomogeneousType)
+            auto sortdata = transposed(data);
+        else
+            auto ref sortdata = data;
```
does that even work? Try `alias sortdata = data;`

Kriyszig replied:
I tested it on the online ide: https://run.dlang.io/is/d8K9sr
Even alias does the job: https://run.dlang.io/is/X9Xx1O

`alias` afaik doesn't allocate the data again so it's indeed a a better option

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -187,25 +188,47 @@ public:
                 }
 
-                static foreach(j; 0 .. RowType.length)
+                static if(isHomogeneousType)
```
create an auxiliary function that does something like 

```
void homogenousDispatch(alias func)
{
     static if(isHomogeneousType)
         func(dataIndex);
    else
        static foreach(j; 0 .. RowType.length)
              if(j == dataIndex)
                     func(dataIndex);
}
```

and  the call it like

```
homogenousDispatch!(
    (j)
    {
          size_t maxsize1 = 0, maxsize2 = 0;
          // ...
    });
```

Ditto throughout.

If you've got a better name for it please use it instead.

Kriyszig replied:
Figured out dispatcher long at last :sweat_smile:

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -187,25 +188,47 @@ public:
                     maxGap = (maxGap > lenCol)? maxGap: lenCol;
```
https://dlang.org/phobos/std_algorithm_comparison.html#.max

Ditto throughout.

thewilsonator commented on source/magpie/group.d:
```diff
@@ -699,24 +729,36 @@ public:
+            static if(kind == Universal)
+                Slice!(Type*, 1, kind) ret = slice!(Type)(elementCountTill[pos[0] + 1] - elementCountTill[pos[0]]).universal;
+            else static if(kind == Canonical)
+                Slice!(Type*, 1, kind) ret = slice!(string)(elementCountTill[pos[0] + 1] - elementCountTill[pos[0]]).canonical;
```
this should either be `Slice!(string*, 1, kind) ret = slice!(string)` or Slice!(Type*, 1, kind) ret = slice!(Type)`

thewilsonator commented on source/magpie/operation.d:
```diff
@@ -203,6 +204,72 @@ auto merge(JoinTypes type = Inner, T, U)(T df1, U df2, string lsuffix = "x_", st
+auto aggregate(int axis, T, Ops...)(T df, Ops ops)
+    if(isDataFrame!T)
+{
+    static if(axis)
```
if calling this with `axis==0` does nothing, it probably makes no sense and should be disallowed.

`auto aggregate(int axis,...)(...) if(isDataFrame!T && axis)

thewilsonator replied:
scratch that axis is unused remove it.

thewilsonator commented on source/magpie/operation.d:
```diff
@@ -203,6 +204,72 @@ auto merge(JoinTypes type = Inner, T, U)(T df1, U df2, string lsuffix = "x_", st
     else static assert(0, "Invalid join type. Available join type: left, right, outer, inner(default)");
 }
 
+enum AggregateOP
```
I would strongly suggest to make this not a restricted set of ops e.g. this misses variance.

`auto aggregate(T, Ops...)(T df)` and then use with

```
auto x = df.aggregate!(min, max);
```

this will simplify the implementation to 
```
static foreach(i;0 ..  Ops.length)
{
    static foreach(j; 0 .. df.RowType.length)
         opres = Ops[i](df.data[j]);
         ret.indx.row.index[0][i] = Ops[i].stringof;
}
```
note that you still do `i` passes through the data, whereas you should be able to do all aggregation statistics with one pass. You should look at [fold](https://dlang.org/phobos/std_algorithm_iteration.html#.fold) and take some inspiration from there, particularly the example `arr.fold!(min, max)`.

thewilsonator commented on source/magpie/operation.d:
```diff
@@ -203,6 +204,72 @@ auto merge(JoinTypes type = Inner, T, U)(T df1, U df2, string lsuffix = "x_", st
+{
+    static if(axis)
+    {
+        DataFrame!(float, df.RowType.length) ret;
```
I'm slightly concerned with the change of data types here, a data frame of ints shouldn't get converted to a df of floats just because I computed the max of a column or row.

thewilsonator commented on source/magpie/operation.d:
```diff
@@ -203,6 +204,72 @@ auto merge(JoinTypes type = Inner, T, U)(T df1, U df2, string lsuffix = "x_", st
+        foreach(i; 0 .. df.RowType.length)
+            ret.data[i].length = Ops.length;
+
+        double opres;
```
ditto.

thewilsonator commented on source/magpie/ops.d:
```diff
@@ -0,0 +1,76 @@
+
+import std.algorithm: map, reduce;
+
+double count(T)(T[] arr)
```
you should try to make aggregate work directly with phobos (and mir's if you can) algorithms directly without the need for this shim.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -188,51 +214,41 @@ public:
-                    {
-                        maxGap = maxsize2;
-                    }
+                        maxsize2 = data[dataIndex][$ - bottom .. $].map!(e => to!string(e).length).reduce!max;   
```
can I suggest making `T[]..map!(e => to!string(e).length).reduce!max` a function and then calling it on the computed ranges.
e.g.

```
/*use a better name*/
auto transform(T)(T[] arr){ return arr.map!(e => to!string(e).length).reduce!max;}
...
if(bottom > data[dataIndex].length)
    maxsize2 = transform(data[dataIndex]);
else if (bottom > 0) 
    maxsize2 = transform(data[dataIndex][$ - bottom .. $]);
```

also computing `to!string(e)` just for the length seems wasteful if you don't use the result. This is totally fine for getting stuff up and running quickly, but you should do some profiling to determine the bottlenecks of this library. 

I suspect GC allocation may be one of them, but do the profiling anyway because it i entirely possible I'm wrong and that the optimiser inlines everything and discards the useless allocation (use LDC for that, DMD won't do much to help with that).

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -895,16 +902,30 @@ public:
-            return data[i2][i1];
-        else
+
+        auto auxDispatch(alias Fn)(size_t indx)
```
no need to declare this multiple times just use the one at the top. If these are on different types move this to you utils file and parameterise on the type.

The whole point was to reduce code remember ;)

thewilsonator replied:
ditto throughout.

Kriyszig replied:
I tried this but it leads to this error:
```
Error: function `magpie.dataframe.DataFrame!(int, 2).DataFrame.auxDispatch!(maxGapCalc).auxDispatch` cannot get frame pointer to magpie.dataframe.DataFrame!(int, 2).DataFrame.display.maxGapCalc!-1L.maxGapCalc
```

The function seems to accessible only within the function scope and couldn't be referenced outside hence I had to bring dispatcher into each function.
Is there a workaround for this?

thewilsonator replied:
Try making auxDispatch `static`.

Kriyszig replied:
making it static gives this:
```
Error: function `magpie.dataframe.DataFrame!(int, 2).DataFrame.display.maxGapCalc!-1L.maxGapCalc` is a nested function and cannot be accessed from magpie.dataframe.DataFrame!(int, 2).DataFrame.auxDispatch!(maxGapCalc).auxDispatch
```
telling nested function cannot be accessed outside.
I fiddled around this for a while however everything led to the same error

thewilsonator replied:
Hmm, maybe try asking on the learn forums. There definitely _should_ be a way to do it.

Kriyszig replied:
I was trying to make a sample example for dlang learn and this worked somehow
https://run.dlang.io/is/EvwVVq
So I was wondering what is the difference between this example and the one I was working with and I noticed if the nested function is a template function, it cannot send a pointer outside.

thewilsonator replied:
I noticed it can if you instantiate it at the call site, i.e. this works (but is not what you want):
```diff
-void nested(int a)
+void nested()(int a)
...
-dispatcher!(nested)(5);
+dispatcher!(nested!())(5);
```

thewilsonator replied:
```
struct S
{
    int data;
    int pos = 5;
private:  
    mixin template auxdispatch(alias F/*, alias indx*/)
    {
        auto auxdispatch(int x /*= indx*/)
        {
            return F(x);
        }
    }
 public:   
    void outer()
    {
        void nested()(int a)
        {
            data += a;
        }
        
        mixin auxdispatch!(nested/*,pos*/);
		auxdispatch(pos);
        import std.stdio;
        writeln(data);
    }
}

void main()
{
    S s;
    s.outer();
}
```

will do the trick, the commented out sections are what I would have liked to have done but a DMD bug prevents it from compiling.

Kriyszig replied:
Thanks a lot for the above example @thewilsonator
Added mixin template in 403aa24

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -945,20 +973,32 @@ public:
-                }
-            }
+
+        void assignAux(ptrdiff_t si = -1)(size_t ri = 0) @property
```
ditto. there is an identical one just above.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -93,16 +149,13 @@ public:
-                        thisGap = tmp;
-                    }
+                    thisGap = max(thisGap, tmp);
                 }
```
You have this same block of code repeated three times with different different indices, you should make that a (nested) function.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -869,13 +962,19 @@ public:
+            static if(si > -1)
+                alias i = si;
+            else
+                size_t i = ri;
```
you could probably refactor this into a mixin (template) too.

thewilsonator replied:
doing this would make the intent of the function much clearer.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -1532,20 +1706,116 @@ public:
+    +/
+    auto aggregate(int axis, Ops...)() @property
+    {
+        static if(axis)
```
There is a lot of similar code in the two branches, try to push this static if deeper into the function.

thewilsonator replied:
in particular do 
```
static if (axis)
{
    import std.meta: staticMap;
    alias Resolve(T) = suitableType!(resolverInternal!(T, Ops));
    alais DFTypes = staticMap!(Resolve, RowType);
}
else 
     alais DFTypes =aggregateType!(Ops)

DataFrame!(true, DFTypes) ret;
init!(axis)(ret);
...
```

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -3418,3 +3688,115 @@ unittest
+    assert(approxEqual(doubledf.data[1][0], 5.6, 1e-8) && approxEqual(doubledf.data[1][1], 8, 1e-8));
+}
+
+private auto customFunc(T)(T[] arr)
```
move this inside the unittest that uses it and make it static

Kriyszig replied:
Done in e30f57a :+1:

thewilsonator commented on source/magpie/group.d:
```diff
@@ -718,6 +829,105 @@ public:
+
+    auto aggregate(int axis, Ops...)() @property
+    {
+        static if(axis)
```
ditto, there is a lot of similar code and it makes it difficult to see what is different.

Kriyszig replied:
I put all the initialisation into nested function :octocat:

thewilsonator commented on source/magpie/group.d:
```diff
@@ -1304,3 +1514,97 @@ unittest
+    assert(approxEqual(dub.data[1][0], 2, 1e-8) && approxEqual(dub.data[1][1], 3, 1e-8) && approxEqual(dub.data[1][2], 5, 1e-8) && approxEqual(dub.data[1][3], 6, 1e-8));
+}
+
+private auto customFunc(T)(T[] arr)
```
ditto, move this inside the unittest.

Kriyszig replied:
Addressed in e30f57a :+1:

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -50,6 +97,17 @@ public:
+        auto gapCalc(T)(T[] arr)
+        {
+            static if(is(T == string))
+                return arr.map!(e => e.length).reduce!max;
```
This static if branch is not needed as `to!string` on a string is a no-op. just use the line below.

thewilsonator commented on source/magpie/group.d:
```diff
@@ -718,6 +829,113 @@ public:
+                        }
+
+            ret.grpIndex.generateCodes();
+            return ret;
```
These two lines are duplicate in each branch.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -1532,20 +1706,122 @@ public:
+            }
+
+            ret.indx.generateCodes();
+            return ret;
```
ditto

thewilsonator commented on source/magpie/group.d:
```diff
@@ -718,6 +829,113 @@ public:
+            alias Resolve(T) = suitableType!(resolverInternal!(T, Ops));
+            alias ResolvedTypes = staticMap!(Resolve, GrpRowType);
+            Group!(ResolvedTypes) ret;
+            init!(axis)(ret);
```
ditto

thewilsonator commented on source/magpie/group.d:
```diff
@@ -17,7 +17,13 @@ struct Group(GrpRowType...)
+    static if(!isHomogeneous!(GrpRowType))
+        alias GrpType = staticMap!(toArr, GrpRowType);
+    else
+        alias GrpType = toArr!(GrpRowType[0])[GrpRowType.length];
```
this code is duplicate with dataframe, move it to utils, and do 
```
alias GrpType = homogeneousTypesFor!(GrpRowType);
```

if you need `isHomogeneousType` use a mixin template.

thewilsonator commented on source/magpie/group.d:
```diff
@@ -93,6 +99,47 @@ private:
+
+            assert(0);
+        }
+    }
```
all of this code is duplicate too.

thewilsonator commented on source/magpie/group.d:
```diff
@@ -528,31 +607,50 @@ public:
+            mixin auxDispatch!(mixinAux);
+            auxDispatch(pos[1]);
         }
         else static if(T.length == 1 && isArray!(T[0]))
```
The only thing different between this branch and the one above is 
```diff
-mixin("data[i][j] " ~ op ~ "= elements.data[j - elementCountTill[pos[0]]];");
+mixin("data[i][j] " ~ op ~ "= elements.data[j - elementCountTill[pos[0]]].get!(GrpRowType[typepos]);");
```
you should merge them.

thewilsonator commented on source/magpie/dataframe.d:
```diff
@@ -1034,15 +1134,28 @@ public:
             }
+            
+            mixin auxDispatch!(assignAux);
+            auxDispatch(pos);
```
the line count increase for doing this here is quite large and the entire point of it was to increase readability which I don't think this does.

thewilsonator commented:
One thing I want you to think about is the amount of similarity and differences between Group and DataFrame and if the differences actually need to be different at all. 

For example they can both be of homogeneous types (or not) and they both contain a field that depend on that property and they both have an index. But, for the data frame the alias for the (in)homogeneous type is `FrameType` in data frame and `GrpType` (the field is both called `data` which is good) and the index is called `indx` for data frames and `grpIndex`.

This means if all I care about the thing I have as a template is `typeof(data)` and the index, I _must_ treat them differently. 

This affect you too. The implementation of `aggregate` for Groups & DataFrames is almost iIdentical, _almost_, but not quite. Now some of this is by necessity, they are different data types after all, but having to implement it twice is not a desirable property of the code. 

In D this is usually done as a free function that then gets called via Uniform Function Call Syntax, that is to say, one writes

```D
struct DataFrame(...)
{
    //...
}

struct Group(...)
{
    //...
}

auto aggregate(T, int axis, Ops...)(ref T t) // T is a dataframe or a Group
{
    //...
}

//User code
void func()
{
    Dataframe!(whatever) df;
    Group!(whatever) grp;
    auto aggdf = df.aggregate!(0,max);
    auto agggrp = grp.aggregate!(0,max);
}
```

I want you to do that, and then for every difference between `DataFrame` and `Group` in the implementation of `aggregate`, ask yourself "does this difference need to exist?".

This separation of behaviour from data is very widely used in D, it is for example the basis of the range: a data structure provides the implementation of how to iterate over it, and then all the algorithms on phobos can be used with that data structure (providing the constraints are met).
 
(Sometimes it is necessary to specialise an algorithm for a data structure because of performance but that is rare, and you should do profiling to make sure that it is worth doing)

Kriyszig commented:
I was on the track of doing exactly that: an aggregate that takes either Group or DataFrame but with the custom function changes, I ran into this trouble:

**https://run.dlang.io/is/88rXa3**

At that point I had implemented aggregate as an overload so I moved each of them into the respective structs and optimized them individually. Some parts are indeed very similar - If the above mentioned problem can be circumvented, I can move the aggregate into a single function by this weekend
